python/social_bot/envs/pr2.py

```diff
@@ -58,128 +81,288 @@ def __init__(self,
         return dist
 
-    def step(self, action):
-        controls = dict(zip(self._r_arm_joints, action))
+    def _get_gripper_pos(self):
+        state = self._agent.get_joint_state("pr2::pr2::r_gripper_joint")
+        return state.get_positions()[0]
+
+    def step(self, actions):
+        def trunc_scale(a, limit):
+            if a < -1.0:
+                return -limit
+            elif a > 1.0:
+                return limit
+            else:
+                return a * limit
+
+        scaled_actions = [trunc_scale(x, self._r_arm_joints_limits[i]) \
+                          for i,x in enumerate(actions)]
+
+        # final actions used to control each joints are:
+        # joint_limit_{i} *  \pi (state)_{i}
+        actions = scaled_actions
+
+        controls = dict(zip(self._r_arm_joints, actions))
 
         self._agent.take_action(controls)
         self._world.step(100)
+
         obs = self._get_observation()
 
         self._steps_in_this_episode += 1
         dist = self._get_finger_tip_distance()
+        gripper_pos = self._get_gripper_pos()
 
         done = self._steps_in_this_episode >= self._max_steps
 
-        reward = -0.01 if (not done) else (1.2 - dist)
+        reward = 0 if (not done) else 1.0 - np.tanh(5.0 * dist)
+
+        dist_reward = self._prev_dist - dist
+
+        pos_reward = self._gripper_reward_dir * (gripper_pos - self._prev_gripper_pos)
 
-        shape_reward = self._prev_dist - dist
         if self._reward_shaping:
-            reward += shape_reward
+            reward += dist_reward + pos_reward
+
+        delta_reward = 0
+
+        if self._l_touch:
+            logger.info("l finger touch!")
+            delta_reward += 0.5
 
-        collision_cnt = self._goal.get_collision_count("beer_contact")
-        if collision_cnt > 1:
-            logger.debug("collide_cnt:" + str(collision_cnt))
+        if self._r_touch:
+            logger.info("r finger touch!")
+            delta_reward += 0.5
 
-        goal_loc, _ = self._goal.get_pose()
+        if self._l_touch and self._r_touch:
+            logger.info("both touch!")
+            goal_loc = self._goal_pose[0]
 
-        goal_move_reward = (goal_loc[2] > 0.001)
-        if goal_move_reward:
-            logger.debug("beer lift! " + str(goal_loc))
-            reward += 10
+            # lifting reward
+            elevation = goal_loc[2] - self._table_height
+            lift = min(max(elevation - 0.01, 0), 0.2)
+
+            if lift > 0:
+                logger.info("beer lift! " + str(lift))
+                delta_reward += (1.0 + 50 * lift)
+
+        if delta_reward > 0:
+            reward += delta_reward
+        else:
+            reward += -0.01     #if no positive reward, penalize it a bit to speed up
+
+        if self._motion_loss > 0.0:
+            v2s = 0.0
+            for joint in self._r_arm_joints:
+                v = self._agent.get_joint_state(joint).get_velocities()[0]
```

Comment:
Maybe we can also simply use actions here. Naturally a human may want to avoid extra efforts (forces of the joints).

Reply:
yes, ||f||_2 could be better here, but we did not use it yet.